### ISSUE TITLE
Add recursive call limit

### ISSUE THREAD

**Issue by hackaugusto** [0xPolygonMiden/miden-vm#1305]
Currently the VM uses recursion to interpret call, which means an attacker can easily write a program that will exhaust the host stack, killing it.

POC: A `dyncall` instruction in MASM causes the Rust program to [recurse](https://github.com/0xPolygonMiden/miden-vm/blob/3a957f7c90176914bda2139f74bff9e5700d59ac/processor/src/lib.rs#L245). The following program can be used to create infinite recursion:

```
proc.rec
  dynexec
end

begin
  procref.rec
  exec.rec
end
```

Running the above causes a crash:

```
thread 'main' has overflowed its stack
fatal runtime error: stack overflow
[1]    99618 abort      cargo run --features="executable" -- run --assembly t.masm
```

**Comment by hackaugusto** [comment]
Another approach is to change the way the bytecode is interpreted, and rely exclusively on the max cycle counter to prevent running out of main memory. This is related to #1217

**Comment by cyberbono3** [comment]
Can I start working on it?

**Comment by hackaugusto** [comment]
> Can I start working on it?

@bobbinth @bitwalker ping. I'm not sure what are the plans with the interpreter rewrite to remove the recursion calls.

**Comment by bobbinth** [comment]
@cyberbono3 - thank you for wanting to help with this! I would probably hold off on this issue for now as we will be refactoring how the processor runs though the program in the next couple of weeks (the `Program` struct will be moved to table-based description of MAST - see https://github.com/0xPolygonMiden/miden-vm/issues/1226). So, the mechanism that may apply to the current implementation may need to be different after the refactor.

**Comment by bobbinth** [comment]
@plafer - is this still relevant in light of the work in the fast processor?

**Comment by plafer** [comment]
No, this is no longer relevant now that we execute the MAST forest iteratively.
